src/ILVerify/src/Verifier.cs

```diff
@@ -0,0 +1,224 @@
+
+        public void SetSystemModuleName(AssemblyName name)
+        {
+            _typeSystemContext.SetSystemModule(_typeSystemContext.GetModule(name));
+        }
+
+        public VerificationResult Verify(AssemblyName moduleToVerify)
+        {
+            if (moduleToVerify == null)
+            {
+                throw new ArgumentNullException(nameof(moduleToVerify));
+            }
+
+            try
+            {
+                if (_typeSystemContext.SystemModule is null)
+                {
+                    return new VerificationResult() { NumErrors = 1, Message = "No system module specified" };
+                }
+
+                EcmaModule module = _typeSystemContext.GetModule(moduleToVerify);
+                return VerifyModule(module, moduleToVerify.Name);
+            }
+            catch (VerifierException e)
+            {
+                return new VerificationResult() { NumErrors = 1, Message = e.Message };
+            }
+        }
+
+        private VerificationResult VerifyModule(EcmaModule module, string path)
+        {
+            foreach (var methodHandle in module.MetadataReader.MethodDefinitions)
+            {
+                var method = (EcmaMethod)module.GetMethod(methodHandle);
+
+                var methodIL = EcmaMethodIL.Create(method);
+                if (methodIL == null)
+                {
+                    continue;
+                }
+
+                var methodName = method.ToString();
+                if (ShouldVerifyMethod != null && !ShouldVerifyMethod(methodName))
+                {
+                    continue;
+                }
+
+                var result = VerifyMethod(method, methodIL, path);
+                if (result.NumErrors > 0)
+                {
+                    return result;
+                }
+            }
+
+            return new VerificationResult();
+        }
+
+        internal VerificationResult VerifyMethod(MethodDesc method, MethodIL methodIL, string moduleName)
+        {
+            StringBuilder output = new StringBuilder();
+            int numErrors = 0;
+            var errors = new List<VerifierError>();
+
+            try
+            {
+                var importer = new ILImporter(method, methodIL);
+
+                importer.ReportVerificationError = (args) =>
+                {
+                    AppendError(method, moduleName, args, output);
+                    errors.Add(args.Code);
+                    numErrors++;
+                };
+
+                importer.Verify();
+            }
+            catch (NotImplementedException e)
+            {
+                output.AppendLine($"Error in {method}: {e.Message}");
+                numErrors++;
+            }
+            catch (InvalidProgramException e)
+            {
+                output.AppendLine($"Error in {method}: {e.Message}");
+                numErrors++;
+            }
+            catch (VerificationException)
+            {
+                numErrors++;
```

Comment:
`VerificationException`s are thrown to either abort the entire method or a single basic block. When such an event happens an error should already have been reported, therefore increasing `numErrors` here is redundant.
It also breaks some test cases.

Reply:
Thanks.
It seems I've not been running all the tests locally and CI isn't either :-( Investing in test automation would surely help.